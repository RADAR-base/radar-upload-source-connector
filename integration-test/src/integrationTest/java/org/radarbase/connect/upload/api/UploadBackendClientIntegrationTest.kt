/*
 *
 *  * Copyright 2019 The Hyve
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  * you may not use this file except in compliance with the License.
 *  * You may obtain a copy of the License at
 *  *
 *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  * See the License for the specific language governing permissions and
 *  * limitations under the License.
 *  *
 *
 */

package org.radarbase.connect.upload.api

import org.apache.kafka.connect.source.SourceRecord
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.*
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.radarbase.connect.upload.converter.ConverterLogRepository
import org.radarbase.connect.upload.converter.LogRepository
import org.radarbase.connect.upload.converter.phone.AccelerometerConverterFactory
import org.radarbase.connect.upload.util.TestBase.Companion.baseUri
import org.radarbase.connect.upload.util.TestBase.Companion.clientCredentialsAuthorizer
import org.radarbase.connect.upload.util.TestBase.Companion.createRecordAndUploadContent
import org.radarbase.connect.upload.util.TestBase.Companion.getAccessToken
import org.radarbase.connect.upload.util.TestBase.Companion.httpClient
import org.radarbase.connect.upload.util.TestBase.Companion.sourceTypeName
import org.radarbase.connect.upload.util.TestBase.Companion.uploadBackendConfig
import org.radarbase.jersey.GrizzlyServer
import org.radarbase.jersey.config.ConfigLoader
import org.radarbase.upload.Config
import org.radarbase.upload.doa.entity.RecordStatus
import org.radarbase.upload.inject.ManagementPortalEnhancerFactory
import java.io.File


@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class UploadBackendClientIntegrationTest {
    private lateinit var uploadBackendClient: UploadBackendClient

    private lateinit var logRepository: LogRepository

    private lateinit var config: Config

    private lateinit var server: GrizzlyServer

    private lateinit var accessToken: String

    private val sourceType = sourceTypeName

    private val fileName = "TEST_ACC.csv"


    @BeforeAll
    fun setUp() {
        uploadBackendClient = UploadBackendClient(
                clientCredentialsAuthorizer,
                httpClient,
                baseUri
        )

        logRepository = ConverterLogRepository()

        accessToken = getAccessToken()

        config = uploadBackendConfig
        val resources = ConfigLoader.loadResources(ManagementPortalEnhancerFactory::class.java, config)

        server = GrizzlyServer(config.baseUri, resources)
        server.start()
    }

    @AfterAll
    fun cleanUp() {
        server.shutdown()
    }

    @Test
    fun requestAllConnectors() {
        val connectors = uploadBackendClient.requestAllConnectors()
        assertNotNull(connectors)
        assertThat(connectors.sourceTypes, not(empty()))
        assertNotNull(connectors.sourceTypes.find { it.name == sourceType })
    }

    @Test
    fun requestConnectorConfigurations() {
        val connectors = uploadBackendClient.requestConnectorConfig(sourceType)
        assertNotNull(connectors)
        assertEquals(sourceType, connectors.name)
    }

    @Test
    fun testRecordCreationToConversionWorkFlow() {
        val createdRecord = createRecordAndUploadContent(accessToken, sourceType, fileName)
        retrieveFile(createdRecord)

        val records = pollRecords()

        val sourceType = uploadBackendClient.requestConnectorConfig(sourceType)

        val converter = AccelerometerConverterFactory()
                .converter(emptyMap(), sourceType, uploadBackendClient, logRepository)

        val recordToProcess = records.records.first { recordDTO -> recordDTO.sourceType == sourceTypeName }
        createdRecord.metadata = uploadBackendClient.updateStatus(recordToProcess.id!!, recordToProcess.metadata!!.copy(status = "PROCESSING", message = "The record is being processed"))
        val convertedRecords = mutableListOf<SourceRecord>()
        converter.convert(records.records.first(), convertedRecords::add)
        assertThat(convertedRecords, not(nullValue()))
        assertThat(convertedRecords, not(empty()))
    }

    private fun pollRecords(): RecordContainerDTO {
        val pollConfig = PollDTO(
                limit = 10,
                supportedConverters = setOf(sourceType))
        val records = uploadBackendClient.pollRecords(pollConfig)
        assertNotNull(records)
        assertThat(records.records.size, greaterThan(0))
        records.records.forEach { recordDTO ->
            assertThat(recordDTO.metadata?.status, equalTo(RecordStatus.QUEUED.toString()))
        }
        println("Polled ${records.records.size} records")
        return records
    }

    private fun retrieveFile(recordId: RecordDTO) {
        uploadBackendClient.retrieveFile(recordId, fileName) { response ->
            assertNotNull(response)
            val responseData = response.bytes()
            assertThat(responseData.size.toLong(), equalTo(File(fileName).length()))
            assertThat(responseData, equalTo(File(fileName).readBytes()))
        }
    }

}
