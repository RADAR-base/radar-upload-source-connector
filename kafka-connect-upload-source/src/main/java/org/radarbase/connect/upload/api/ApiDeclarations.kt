/*
 *
 *  * Copyright 2019 The Hyve
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  * you may not use this file except in compliance with the License.
 *  * You may obtain a copy of the License at
 *  *
 *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  * See the License for the specific language governing permissions and
 *  * limitations under the License.
 *  *
 *
 */

package org.radarbase.connect.upload.api

import com.fasterxml.jackson.annotation.JsonProperty
import java.time.Instant
import java.time.LocalDateTime

enum class LogLevel {
    INFO, DEBUG, WARN, ERROR}

data class Log (
        var logLevel: LogLevel,
        var message: String)

data class LogsDto(
        var url: String? = null,
        var contents: String? = null)

data class ContentsDTO(
        var url: String? = null,
        var text: String? = null,
        var contentType: String,
        var createdDate: Instant,
        var size: Long,
        var fileName: String)

data class RecordDataDTO(
        var projectId: String?,
        var userId: String?,
        var sourceId: String?,
        var time: LocalDateTime? = null,
        var timeZoneOffset: Int? = null,
        var contents: Set<ContentsDTO>? = null)

data class RecordMetadataDTO(
        var revision: Int,
        var status: String,
        var message: String? = null,
        var createdDate: Instant? = null,
        var modifiedDate: Instant? = null,
        var committedDate: Instant? = null,
        var logs: LogsDto? = null)


data class RecordDTO(
        var id: Long?,
        var data: RecordDataDTO?,
        var sourceType: String?,
        var metadata: RecordMetadataDTO?)

data class RecordContainerDTO(
        var limit: Int? = null,
        var records: List<RecordDTO>)

data class SourceTypeDTO(
        var name: String?,
        var topics: Set<String>?,
        var contentTypes: Set<String>?,
        var timeRequired: Boolean?,
        var sourceIdRequired: Boolean?,
        var configuration: Map<String, String>?
)

data class SourceTypeContainerDTO(
        var sourceTypes: List<SourceTypeDTO>)

data class PollDTO(
        var limit: Int = 10,
        var supportedConverters: List<String>)


data class OauthToken(
        @JsonProperty("access_token") var accessToken: String,
        @JsonProperty("expires_in") var expiresIn: Long,
        @JsonProperty("iat") var issuedAt: Long,
        @JsonProperty("scope") var scope: String,
        @JsonProperty("sub") var sub: String)
