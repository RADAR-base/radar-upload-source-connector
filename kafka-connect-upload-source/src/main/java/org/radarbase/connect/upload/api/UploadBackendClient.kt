/*
 *
 *  * Copyright 2019 The Hyve
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  * you may not use this file except in compliance with the License.
 *  * You may obtain a copy of the License at
 *  *
 *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  * See the License for the specific language governing permissions and
 *  * limitations under the License.
 *  *
 *
 */

package org.radarbase.connect.upload.api

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import com.fasterxml.jackson.module.kotlin.jsonMapper
import com.fasterxml.jackson.module.kotlin.kotlinModule
import okhttp3.Interceptor
import okhttp3.MediaType
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.ResponseBody
import okio.BufferedSink
import org.radarbase.connect.upload.exception.BadGatewayException
import org.radarbase.connect.upload.exception.ConflictException
import org.radarbase.connect.upload.exception.NotAuthorizedException
import org.radarbase.connect.upload.logging.Log
import org.slf4j.LoggerFactory
import java.io.Closeable
import java.io.IOException

open class UploadBackendClient(
    authInterceptor: Interceptor,
    private var httpClient: OkHttpClient,
    private var uploadBackendBaseUrl: String,
) : Closeable {

    init {
        httpClient = httpClient
            .newBuilder()
            .addInterceptor(authInterceptor)
            .build()

        uploadBackendBaseUrl = uploadBackendBaseUrl.trimEnd('/')
    }

    open fun pollRecords(configuration: PollDTO): RecordContainerDTO = httpClient.executeRequest {
        url("$uploadBackendBaseUrl/records/poll")
        post(configuration.toJsonBody())
    }

    open fun requestConnectorConfig(name: String): SourceTypeDTO = httpClient.executeRequest {
        url("$uploadBackendBaseUrl/source-types/$name")
    }

    open fun requestAllConnectors(): SourceTypeContainerDTO = httpClient.executeRequest {
        url("$uploadBackendBaseUrl/source-types")
    }

    open fun <T> retrieveFile(
        record: RecordDTO,
        fileName: String,
        handling: (ResponseBody) -> T,
    ): T {
        return httpClient.executeRequest({
            url("$uploadBackendBaseUrl/records/${record.id}/contents/$fileName")
        }) {
            handling(it.body ?: throw IOException("No file content response body"))
        }
    }

    open fun retrieveRecordMetadata(recordId: Long): RecordMetadataDTO = httpClient.executeRequest {
        url("$uploadBackendBaseUrl/records/$recordId/metadata")
    }

    open fun updateStatus(recordId: Long, newStatus: RecordMetadataDTO): RecordMetadataDTO {
        val result: RecordMetadataDTO = httpClient.executeRequest {
            url("$uploadBackendBaseUrl/records/$recordId/metadata")
            post(newStatus.toJsonBody())
        }
        logger.info(
            "Successfully updated record {} status to {} (rev. {})",
            recordId,
            result.status,
            result.revision,
        )
        return result
    }

    open fun addLogs(log: Log): RecordMetadataDTO = httpClient.executeRequest {
        url("$uploadBackendBaseUrl/records/${log.recordId}/logs")
        put(object : RequestBody() {
            override fun contentType() = TEXT_PLAIN

            override fun writeTo(sink: BufferedSink) = log.asString(sink)
        })
    }

    override fun close() {
    }

    private inline fun <reified T : Any> OkHttpClient.executeRequest(
        noinline requestBuilder: Request.Builder.() -> Request.Builder,
    ): T = executeRequest(requestBuilder) { response ->
        mapper.readValue(response.body?.byteStream(), T::class.java)
            ?: throw IOException("Received invalid response")
    }

    private fun <T> OkHttpClient.executeRequest(
        requestBuilder: Request.Builder.() -> Request.Builder,
        handling: (Response) -> T,
    ): T {
        val request = Request.Builder().requestBuilder().build()
        return this.newCall(request).execute().use { response ->
            if (response.isSuccessful) {
                logger.info("Request to ${request.url} is SUCCESSFUL")
                handling(response)
            } else {
                logger.info("Request to ${request.url} has FAILED with response-code ${response.code}")
                when (response.code) {
                    401 -> throw NotAuthorizedException("access token is not provided or is invalid : ${response.message}")
                    403 -> throw NotAuthorizedException("access token is not authorized to perform this request")
                    409 -> throw ConflictException("Conflicting request exception: ${response.message}")
                    else -> throw BadGatewayException("Failed to make request to ${request.url}: Error code ${response.code}:  ${response.body?.string()}")
                }
            }
        }
    }

    private fun Any.toJsonBody(mediaType: MediaType = APPLICATION_JSON): RequestBody = mapper
        .writeValueAsBytes(this)
        .toRequestBody(mediaType)

    companion object {
        private val logger = LoggerFactory.getLogger(UploadBackendClient::class.java)
        private val APPLICATION_JSON = "application/json; charset=utf-8".toMediaType()
        private val TEXT_PLAIN = "text/plain; charset=utf-8".toMediaType()
        private var mapper: ObjectMapper = jsonMapper {
            addModule(kotlinModule())
            addModule(JavaTimeModule())
            disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
            disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
        }
    }
}
