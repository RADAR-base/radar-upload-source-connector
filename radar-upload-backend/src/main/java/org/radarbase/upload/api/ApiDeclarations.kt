/*
 *
 *  * Copyright 2019 The Hyve
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  * you may not use this file except in compliance with the License.
 *  * You may obtain a copy of the License at
 *  *
 *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  * See the License for the specific language governing permissions and
 *  * limitations under the License.
 *  *
 *
 */

package org.radarbase.upload.api

import java.time.Instant
import java.time.LocalDateTime

data class LogsDto(
        var url: String? = null,
        var contents: String? = null)

data class ContentsDTO(
        var url: String? = null,
        var text: String? = null,
        var contentType: String,
        var createdDate: Instant? = null,
        var size: Long? = null,
        var fileName: String)

data class RecordDataDTO(
        var projectId: String?,
        var userId: String?,
        var sourceId: String?,
        var time: LocalDateTime? = null,
        var timeZoneOffset: Int? = null,
        var contents: Set<ContentsDTO>? = null)

data class RecordMetadataDTO(
        var id: Long? = null,
        var revision: Int? = null,
        var status: String? = null,
        var message: String? = null,
        var createdDate: Instant? = null,
        var modifiedDate: Instant? = null,
        var committedDate: Instant? = null,
        var logs: LogsDto? = null,
        var callbackUrl: String? = null)

data class RecordDTO(
        var id: Long?,
        var data: RecordDataDTO?,
        var sourceType: String?,
        var metadata: RecordMetadataDTO?)

data class RecordContainerDTO(
        var limit: Int? = null,
        var records: List<RecordDTO>,
        var lastId: Long? = null)

data class SourceTypeDTO(
        var name: String,
        var topics: Set<String>?,
        var contentTypes: Set<String>?,
        var timeRequired: Boolean?,
        var sourceIdRequired: Boolean?,
        var configuration: Map<String, String>?
)

data class SourceTypeContainerDTO(
        var sourceTypes: List<SourceTypeDTO>)

data class PollDTO(
        var limit: Int = 10,
        var supportedConverters: List<String>)
