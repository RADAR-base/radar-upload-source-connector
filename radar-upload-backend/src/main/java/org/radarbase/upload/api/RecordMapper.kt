/*
 *
 *  * Copyright 2019 The Hyve
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  * you may not use this file except in compliance with the License.
 *  * You may obtain a copy of the License at
 *  *
 *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  * See the License for the specific language governing permissions and
 *  * limitations under the License.
 *  *
 *
 */

package org.radarbase.upload.api

import org.radarbase.upload.doa.entity.Record
import org.radarbase.upload.doa.entity.RecordContent
import org.radarbase.upload.doa.entity.RecordMetadata

interface RecordMapper {
    suspend fun cleanBaseUri(): String

    suspend fun fromRecord(record: Record): RecordDTO
    suspend fun fromRecords(records: List<Record>, page: Page?): RecordContainerDTO
    suspend fun fromMetadata(metadata: RecordMetadata): RecordMetadataDTO
    suspend fun fromContent(content: RecordContent): ContentsDTO

    suspend fun toRecord(record: RecordDTO): Pair<Record, RecordMetadata>
}
