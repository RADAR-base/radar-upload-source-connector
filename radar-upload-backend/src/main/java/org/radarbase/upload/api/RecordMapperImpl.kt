/*
 *
 *  * Copyright 2019 The Hyve
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  * you may not use this file except in compliance with the License.
 *  * You may obtain a copy of the License at
 *  *
 *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  * See the License for the specific language governing permissions and
 *  * limitations under the License.
 *  *
 *
 */

package org.radarbase.upload.api

import org.radarbase.upload.Config
import org.radarbase.upload.doa.SourceTypeRepository
import org.radarbase.upload.doa.entity.Record
import org.radarbase.upload.doa.entity.RecordContent
import org.radarbase.upload.doa.entity.RecordMetadata
import org.radarbase.upload.doa.entity.RecordStatus
import java.lang.IllegalArgumentException
import java.net.URLEncoder
import java.time.Instant
import javax.ws.rs.BadRequestException
import javax.ws.rs.core.Context
import javax.ws.rs.core.UriInfo

class RecordMapperImpl(
        @Context val uri: UriInfo,
        @Context val sourceTypeRepository: SourceTypeRepository,
        @Context val config: Config) : RecordMapper {

    private val cleanBaseUri: String
        get() = (config.advertisedBaseUri ?: uri.baseUri).toString().trimEnd('/')

    override fun toRecord(record: RecordDTO): Pair<Record, RecordMetadata> {
        val recordDoa = Record().apply {
            val data = record.data ?: throw BadRequestException("No data field included")
            projectId = data.projectId ?: throw BadRequestException("Missing project ID")
            userId = data.userId ?: throw BadRequestException("Missing user ID")
            sourceId = data.sourceId ?: throw BadRequestException("Missing source ID")
            sourceType = sourceTypeRepository.read(record.sourceType
                    ?: throw BadRequestException("Missing source type"))
                    ?: throw BadRequestException("Source type not found")
        }

        return Pair(recordDoa, toMetadata(record.metadata))
    }

    private fun toMetadata(metadata: RecordMetadataDTO?) = RecordMetadata().apply {
        createdDate = Instant.now()
        modifiedDate = Instant.now()
        revision = 1

        status = metadata?.status?.let {
            try {
                RecordStatus.valueOf(it)
            } catch (ex: IllegalArgumentException) {
                null
            }
        } ?: RecordStatus.INCOMPLETE

        callbackUrl = metadata?.callbackUrl
    }

    override fun fromRecord(record: Record) = RecordDTO(
            id = record.id!!,
            metadata = fromMetadata(record.metadata).apply { id = null },
            sourceType = record.sourceType.name,
            data = RecordDataDTO(
                    projectId = record.projectId,
                    userId = record.userId,
                    sourceId = record.sourceId,
                    time = record.time,
                    timeZoneOffset = record.timeZoneOffset,
                    contents = record.contents?.mapTo(HashSet(), this::fromContent)
            ))

    override fun fromRecords(records: List<Record>, limit: Int) = RecordContainerDTO(
            limit = limit,
            records = records.map(::fromRecord),
            lastId = records.maxBy { record ->  record.id!! }?.id!!)

    override fun fromContent(content: RecordContent): ContentsDTO {
        val cleanedFileName = URLEncoder.encode(content.fileName, "UTF-8")
                .replace("+", "%20")

        return ContentsDTO(
                url = "$cleanBaseUri/records/${content.record.id}/contents/$cleanedFileName",
                contentType = content.contentType,
                createdDate = content.createdDate,
                size = content.size,
                fileName = content.fileName)
    }

    override fun fromMetadata(metadata: RecordMetadata) = RecordMetadataDTO(
            id = metadata.id,
            revision = metadata.revision,
            status = metadata.status.name,
            message = metadata.message,
            createdDate = metadata.createdDate,
            modifiedDate = metadata.modifiedDate,
            committedDate = metadata.committedDate,
            // use record.id, since metadata and record have one-to-one
            logs = metadata.logs?.let {
                LogsDto(url = "${cleanBaseUri}/records/${metadata.id}/logs")
            }
    )
}
