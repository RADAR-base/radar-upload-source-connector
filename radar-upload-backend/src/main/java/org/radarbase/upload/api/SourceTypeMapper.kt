package org.radarbase.upload.api

import org.radarbase.upload.doa.entity.SourceType

interface SourceTypeMapper {
    fun fromSourceType(record: SourceType): SourceTypeDTO
    fun fromSourceTypes(records: List<SourceType>): SourceTypeContainerDTO
}
