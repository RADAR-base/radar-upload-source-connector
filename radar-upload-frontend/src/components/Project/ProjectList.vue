<template>
  <v-list shaped>
    <v-layout justify-center>
      <v-progress-circular
        class="mt-2"
        v-show="loading"
        indeterminate
        color="primary"
      />
      <v-alert
        type="error"
        v-text="errorMessage"
        v-show="errorMessage"
        dense
        text
      />
    </v-layout>
    <v-list-item-group
      color="primary"
      v-model="selectedProject"
    >
      <v-list-item
        v-for="(project, i) in projects"
        :key="i"
        @click.native="selectProject(project)"
      >
        <v-list-item-icon>
          <v-icon
            color="info"
          >
            mdi-folder-text-outline
          </v-icon>
        </v-list-item-icon>
        <v-list-item-content>
          <v-list-item-title>
            {{ project.text }}
          </v-list-item-title>
        </v-list-item-content>
      </v-list-item>
    </v-list-item-group>
  </v-list>
</template>

<script>
import projectMixin from './projectMixin.js';

export default {
  mixins: [projectMixin],
  methods: {
    selectProjectFromRoute() {
      const { projectId } = this.$route.params;
      if (projectId) {
        const selectedProject = this.projects.find(project => project.value === projectId);
        if (selectedProject) {
          this.selectProject(selectedProject);
          return;
        }
        this.$error('Selected project not found, please select a project to continue');
        this.$router.push('/projects');
      } else {
        this.$router.push('/projects');
      }
    },
  },
  watch: {
    projects: {
      // right after projects list is loaded, check if a project is selected
      handler(projects) {
        if (projects.length > 0 && !this.selectedProject) {
          this.selectProjectFromRoute();
        }
      },
      immediate: true,
    },
    $route: {
      // after component is rendered
      // in case route (projectId in route) is modified manually
      deep: true,
      handler({ params }) {
        const { projectId } = params;
        if (projectId && (projectId !== this.selectedProject)) {
          this.selectProjectFromRoute();
        }
      },
    },
  },
};
</script>

<style>

</style>
