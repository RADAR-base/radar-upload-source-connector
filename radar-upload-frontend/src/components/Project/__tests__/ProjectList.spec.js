/* eslint-disable no-undef */
import { shallowMount } from '@vue/test-utils';
import { Store } from 'vuex-mock-store';
import flushPromises from 'flush-promises';
import ProjectList from '../ProjectList.vue';
import api from '@/axios/project.js';

const projectId = 'id1';
const projectName = 'name1';

api.getProjects = jest.fn().mockResolvedValue([{
  id: projectId,
  name: projectName,
  location: 'Utrecht',
  organization: 'The Hyve',
  description: 'my test project',
}]);


describe('ProjectList', () => {
  const $store = new Store();
  afterEach(() => {
    api.getProjects.mockClear();
  });


  const wrapper = shallowMount(ProjectList, {
    mocks: {
      $store,
      $route: {
        params: {
          projectId: '',
        },
      },
      $router: {
        push: jest.fn(),
      },
      $error: jest.fn(),
    },
    stubs: ['v-list',
      'v-list-item-group',
      'v-list-item',
      'v-list-item-icon',
      'v-icon',
      'v-list-item-content',
      'v-list-item-title',
      'v-progress-circular',
      'v-layout',
      'v-alert',
    ],
    propsData: {},
  });

  it('match snapShot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('watch: projects', () => {
    const selectProjectFromRoute = jest.spyOn(wrapper.vm, 'selectProjectFromRoute');
    wrapper.setData({ selectedProject: '12', projects: [1, 2] });
    expect(selectProjectFromRoute).not.toBeCalled();

    wrapper.setData({ selectedProject: '', projects: [1, 2, 3] });
    expect(selectProjectFromRoute).toBeCalled();
    selectProjectFromRoute.mockClear();
  });

  it('watch: $route', async () => {
    const selectProjectFromRoute = jest.spyOn(wrapper.vm, 'selectProjectFromRoute');
    wrapper.setData({ selectedProject: '12' });
    wrapper.setData({ $route: { params: { projectId: '12' } } });
    expect(selectProjectFromRoute).not.toBeCalled();

    await flushPromises();
    wrapper.setData({ selectedProject: '23' });
    wrapper.setData({ $route: { params: { projectId: 'randomNumber' } } });

    expect(selectProjectFromRoute).toBeCalled();
    selectProjectFromRoute.mockClear();
  });

  it('selectProjectFromRoute', async () => {
    wrapper.setMethods({ selectProject: jest.fn() });

    const selectedProject = { value: 'projectId' };
    wrapper.setData({ projects: [selectedProject] });

    // projectId in route different from that in projects
    wrapper.setData({ $route: { params: { projectId: 'differentPRojectId' } } });
    wrapper.vm.selectProjectFromRoute();
    expect(wrapper.vm.selectProject).not.toBeCalledWith(selectedProject);

    // projectId in route found in projects
    wrapper.setData({ $route: { params: { projectId: 'projectId' } } });
    wrapper.vm.selectProjectFromRoute();
    expect(wrapper.vm.selectProject).toBeCalledWith(selectedProject);
  });
});
