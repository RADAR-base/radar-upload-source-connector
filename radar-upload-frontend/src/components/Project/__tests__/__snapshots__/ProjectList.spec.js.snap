// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectList match snapShot 1`] = `
<v-list-stub shaped="">
  <v-layout-stub justify-center="">
    <v-progress-circular-stub indeterminate="" color="primary" class="mt-2" style="display: none;"></v-progress-circular-stub>
    <v-alert-stub type="error" dense="" text="" style="display: none;"></v-alert-stub>
  </v-layout-stub>
  <v-list-item-group-stub color="primary" value="">
    <v-list-item-stub>
      <v-list-item-icon-stub>
        <v-icon-stub color="info">
          mdi-folder-text-outline
        </v-icon-stub>
      </v-list-item-icon-stub>
      <v-list-item-content-stub>
        <v-list-item-title-stub>

        </v-list-item-title-stub>
      </v-list-item-content-stub>
    </v-list-item-stub>
  </v-list-item-group-stub>
</v-list-stub>
`;
